### DISCUSSION TITLE
Work with a specific Ultralisp distribution?

### DISCUSSION THREAD

**Discussion by mmontone** [ultralisp/ultralisp#285] [Q&A] [answered]
Hi. Is it possible to work with a specific distribution of Ultralisp? Like, point my Quicklisp client to an Ultralisp build with some id from X date.

**Comment by svetlyak40wt** [comment, marked as answer]
Sure, Ultralisp dists are vesioned, but version number includes a time. If you look at a metadata like this, you will see these version numbers:

```
curl https://dist.ultralisp.org/
name: ultralisp
version: 20241010132000
distinfo-subscription-url: http://dist.ultralisp.org/ultralisp.txt
distinfo-template-url: http://dist.ultralisp.org/ultralisp/{{version}}/distinfo.txt
release-index-url: http://dist.ultralisp.org/ultralisp/20241010132000/releases.txt
system-index-url: http://dist.ultralisp.org/ultralisp/20241010132000/systems.txt
canonical-distinfo-url: http://dist.ultralisp.org/ultralisp/20241010132000/distinfo.txt
```

To pin some version, you might use URL like http://dist.ultralisp.org/ultralisp/20241010132000/distinfo.txt.

However updating this version number manually is cumbersome. I prefer to use Qlot. It can move to the latest version number automatically and `qlot.lock` file with the pinned dist version can be commited to the git repository.

**Reply by mmontone** [reply]
Nice. And how can I get the version numbers available? The `20241010132000` ?
Does this info about versions appear in the docs? Could be interesting to have there ...

**Reply by svetlyak40wt** [reply]
There is no a special interface to list all versions of the dist in the UI. If you describe a "User Story", then we might think how to implement such interface.

Also, there is no special quicklisp-versions.txt file in metadata. Using this file you might retrieve versions list in the REPL:

```
CL-USER> (ql:available-dist-versions "ultralisp")
Downloading http://dist.ultralisp.org/ultralisp-versions.txt
; Debugger entered on #<UIOP/RUN-PROGRAM:SUBPROCESS-ERROR {7006EC9523}>
[1] CL-USER> 
; Evaluation aborted on #<UIOP/RUN-PROGRAM:SUBPROCESS-ERROR {7006EC9523}>
CL-USER> (ql:available-dist-versions "quicklisp")
Downloading http://beta.quicklisp.org/dist/quicklisp-versions.txt
##########################################################################
(("2023-10-21"
  . "http://beta.quicklisp.org/dist/quicklisp/2023-10-21/distinfo.txt")
 ("2023-09-06"
  . "http://beta.quicklisp.org/dist/quicklisp/2023-09-06/distinfo.txt")
 ("2023-06-18"
  . "http://beta.quicklisp.org/dist/quicklisp/2023-06-18/distinfo.txt")
 ("2023-02-15"
  . "http://beta.quicklisp.org/dist/quicklisp/2023-02-15/distinfo.txt")
…
```

However this file works for the Quicklisp because there aren't many versions. The file is 9.5Kb size and only 128 lines lenght, where as Ultralisp has more than 13000 versions. I didn't add this file into the metadata because file with such big number of entries will be useless.

Which use case are you trying to solve?

**Reply by mmontone** [reply]
I understand. Thank you for the info. It is useful.

My use case is that our projects used to use Quicklisp, and it was nice while it was being updated with some frequency, but it has not been updated for a year now. Quicklisp being more stable because of its build step was a good thing, and while Ultralisp has more projects, it lives on the edge. And so pinning to a particular Ultralisp dist version and perform a local build could act as a replacement of Quicklisp. 
I'm aware of qlot and other projects, but Quicklisp with a monthly dist build was a good option.

**Comment by mmontone** [comment]
> I prefer to use Qlot. It can move to the latest version number automatically and qlot.lock file with the pinned dist version can be commited to the git repository.

You make Qlot work with Ultralisp? or I am misunderstanding?

**Reply by svetlyak40wt** [reply]
Qlot can work with a multiuple additional quicklisp compatible distributions.

Here is an example of a simple project of mine, which uses Qlot. In the `qlfile` I [say](https://github.com/40ants/github-actions/blob/master/qlfile#L1) I'd like to use Ultralisp. And in `qifile.lock` a particular version of Ultralisp dist [is pinned](https://github.com/40ants/github-actions/blob/master/qlfile.lock#L8). 

When I feel I'm ready to switch to a fresher version of the dist, I do `qlot update` and it changes `qlfile.lock` to pin the latest version of the dist. After that, I can run tests, deploy service into the testing environement. If something goes wrong, I might rollback `qlfile.lock` and search for the problem without a rush.

**Reply by mmontone** [reply]
Oh. That is very good. It is good to know. Thank you.
